resolwe/storage/manager.py

```diff
@@ -18,251 +18,307 @@
-            )[-1]
-        return None
-
-    def _should_delete(self, storage_location: StorageLocation) -> bool:
-        """Get if the given StorageLocation should be deleted.
-
-        :return: True if storage_location should be deleted, False otherwise.
-        :rtype: bool
-        """
-        rules = self._get_rules(storage_location.connector_name, "delete")
-        rule_results = {}
-        rule_results["has_rules"] = len(rules) > 0
 
-        rule_results["not_locked"] = not storage_location.locked
+        default_connector_whens = [
+            models.When(
+                connector_name=connector_name,
+                then=connectors[connector_name].priority,
+            )
+            for connector_name in connectors
+        ]
 
-        # Never remove the last storage location.
         min_other_copies = rules.get("min_other_copies", 1)
-        current_copies = self.file_storage.storage_locations.filter(
-            status=StorageLocation.STATUS_DONE
-        ).count()
-        rule_results["has_copies"] = min_other_copies < current_copies
-
-        # Never remove the location with "highest" priority that is
-        # default_storage_location on FileStorage object.
-        default_location = storage_location.file_storage.default_storage_location
-        rule_results["highest_priority"] = storage_location != default_location
-
-        # Do not delete location with status STATUS_DELETING. Probably error
-        # occured while removing it and we do not want to fall into endless
-        # loop of trying to remove same location over and over again.
-        # The Cleanup Manager will remove it eventually.
-        rule_results["status_ok"] = (
-            storage_location.status == StorageLocation.STATUS_DONE
+        default_delay = self._preprocess_delay(rules.get("delay"))
+
+        return (
+            FileStorage.objects.filter(
+                data__status=Data.STATUS_DONE,
+                created__lt=now()  # type: ignore
+                - models.Case(
+                    *whens,
+                    default=timedelta(days=default_delay),
+                    output_field=models.DurationField(),
+                ),
+            )
+            .annotate(
+                all_locations_count=models.Count(
+                    "storage_locations",
+                    filter=models.Q(
+                        storage_locations__status=StorageLocation.STATUS_DONE
+                    ),
+                    distinct=True,
+                )
+            )
+            .annotate(
+                open_logs=models.Exists(
+                    AccessLog.objects.filter(
+                        storage_location__file_storage=models.OuterRef("id"),
+                        storage_location__connector_name=connector_name,
+                        finished__isnull=True,
+                    )
+                )
+            )
+            .annotate(
+                transfered_lately=models.Exists(
```

Comment:
Do you need this filter, or do `open_logs` and `accessed_lately` cover this use-case? If yes, you have to take care that `last_update` is set before the log is released.

Reply:
Fixed.